ts/client/storageEngine/transactions/db.ts

```diff
@@ -0,0 +1,197 @@
+        }
+        return count;
+    }
+
+    private async throwIfAlreadyExists(txnMessage: string) {
+        try {
+            const itemFound = await this.get(txnMessage);
+            if (itemFound) {
+                throw new TransactionAlreadyExists(txnMessage);
+            }
+        } catch (error) {
+            if (error.name !== "NotFoundError") {
+                throw error;
+            }
+        }
+    }
+
+    private getMessage(msgOrTxn: TransationMessageOrObject): string {
+        if (typeof msgOrTxn === "string") {
+            return msgOrTxn;
+        }
+        return msgOrTxn.message();
+    }
+
+    private async getStatusOrFail(message: string): Promise<TransactionStatus> {
+        try {
+            return await this.get(message);
+        } catch (error) {
+            if (error.name === "NotFoundError") {
+                throw new TransactionDoesNotExist(message);
+            } else {
+                throw error;
+            }
+        }
+    }
+
+    private validateStatusTransition(cur: Status, next: Status) {
+        // Validate transition from pending
+        if (
+            cur === Status.Pending &&
+            (next === Status.Submitted || next === Status.Failed)
+        ) {
+            return;
+            // Validate transition from submited
+        } else if (
+            cur === Status.Submitted &&
+            (next === Status.Finalized || next === Status.Failed)
+        ) {
+            return;
+        }
+        // Fail on everything else
+        throw new StatusTransitionInvalid(cur, next);
+    }
+
+    private async transition(
+        message: string,
+        next: Status,
+        detail?: string
+    ): Promise<TransactionStatus> {
+        const txnStatus = await this.getStatusOrFail(message);
+        this.validateStatusTransition(txnStatus.status, next);
+        const newTxnStatus = {
+            ...txnStatus,
+            status: next,
+            detail
+        };
+        await this.toDB(message, newTxnStatus);
+        return newTxnStatus;
+    }
+
+    private async toDB(msg: string, status: TransactionStatus) {
+        const serialized = JSON.stringify({
+            tx: status.transaction.serialize(),
+            status: status.status,
+            detail: status.detail,
+            batchID: status.batchID,
+            l1TxnHash: status.l1TxnHash,
+            l1BlockIncluded: status.l1BlockIncluded
+        });
+        await this.db.put(msg, serialized);
+    }
+
+    private async fromDB(msg: string): Promise<TransactionStatus> {
+        const object = JSON.parse(await this.db.get(msg));
+
+        return {
+            transaction: TransferOffchainTx.deserialize(arrayify(object.tx)),
+            status: object.status,
+            detail: object.detail || undefined,
+            batchID: object.batchID || undefined,
+            l1TxnHash: object.l1TxnHash || undefined,
+            l1BlockIncluded: object.l1BlockIncluded || undefined
```

Comment:
Are the `|| undefined`'s needed? Is there something else that needs for the field not to be `null` or some other falsy value?

Reply:
Ah right, these values will be undefined implicitly anyway. No need of `|| undefined`